### PR TITLE
Implement Exponential Family

### PR DESCRIPTION
**Pull request by vishwakftw**
Bare bones implementation is ready. More changes will be pushed after testing locally. 
This closes https://github.com/probtorch/pytorch/issues/96.

cc: @fritzo 

Checklist for Exponential Family with closed form solutions for entropy as a function of natural parameters.
- [x] Tests for `.entropy()`
- [x] Tests for `kl_divergence(-,-)`
- [x] Normal
- [x] Exponential
- [x] Gamma
- [x] Bernoulli
- [x] Pareto
- [x] Binomial
- [x] Beta
- [x] Dirichlet
- [ ] Laplace (Is this possible?)

Others can be found in [this table](https://en.wikipedia.org/wiki/Exponential_family#Table_of_distributions).

### CODE REVIEW

**Comment by fritzo** [comment]
Looks great. I think the next step is to add a generic test to ensure `Dist.entropy()` agrees with `ExponentialFamily.entropy()` for all subclasses `Dist` of `ExponentialFamily`. You should be able to fork and modify the existing `TestKL.test_entropy_monte_carlo` for this, let me know if that makes sense.

**Comment by vishwakftw** [comment]
@fritzo One question: since we already have the log normalizer function, finding the derivatives w.r.t. the natural parameters should not be hard / non-trivial. Why bother the AD system then?

**Comment by fritzo** [comment]
> Why bother with the AD system?

I can give two weak reasons: First, the `ExponentialFamily` distribution mainly serves as a test for our highly hand-tuned implementations of `.entropy()` and `.kl_divergence()`. As such we should maximize auditability by performing as little math by hand as possible: we should offload as much as possible to the AD system. In summary: nobody has time to check the math, and that's the point of using standard formulae.

Second, we can probably check correctness of `.log_normalizer` by evaluating `log_prob` in two different ways. Do you think this is possible?

**Comment by fritzo** [comment]
There's no point in implementing this for more distributions until we integrate it with our unit tests. Implementing these things without tests is like writing mathematical definitions without theorems.

**Comment by vishwakftw** [comment]
I have added the tests. However these tests fail, and I unfortunately can't debug the mistake. @fritzo

File: torch/distributions/exp_family.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import torch
+from torch.distributions.distribution import Distribution
+from torch.autograd import Variable
+
+
+class ExponentialFamily(Distribution):
+
+    _zero_carrier_measure = None
+
+    def natural_params(self):
```
Comment (fritzo) [inline torch/distributions/exp_family.py]: It looks like you could simplify code by making this a property:
```py
class ExponentialFamily(Distribution):
    @property
    def natural_params(self):
        raise NotImplementedError

class Bernoulli(ExponentialFamily):
    @lazy_property
    def natural_params(self):
        return ...some math...
```

File: torch/distributions/exp_family.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import torch
+from torch.distributions.distribution import Distribution
+from torch.autograd import Variable
+
+
+class ExponentialFamily(Distribution):
+
+    _zero_carrier_measure = None
```
Comment (fritzo) [inline torch/distributions/exp_family.py]: As I understand from the [Bregman divergence paper](https://pdfs.semanticscholar.org/fab3/337bcac9d026bb0e370cc644c46823e574c9.pdf), entropy can be computed if the expected value of the carrier measure is known (`b` in the paper's notation). Could you replace the `_zero_carrier_measure` attribute with an abstract function or property?
```py
class ExponentialFamily(Distribution):
    @property
    def carrier_measure_mean(self):
        raise NotImplementedError

class Normal(ExponentialFamily):
    carrier_measure_mean = 0
```

File: torch/distributions/exponential.py [hunk 3]
```diff
@@ -39,3 +41,18 @@ def log_prob(self, value):
 
     def entropy(self):
         return 1.0 - torch.log(self.rate)
+
+    def natural_params(self):
+        return self._natural_params
+
+    @lazy_property
+    def _natural_params(self):
+        if isinstance(self.rate, Variable):
```
Comment (fritzo) [inline torch/distributions/exponential.py]: Everything will be a variable soon. Once you merge in the latest master I believe you can assume `isinstance(-, Variable)` in all your implementations.

**Comment by vishwakftw** [comment]
Most of the tests are fine, except 2
1. The first error happens because an input to `Normal` is a `Tensor`. I thought that the inputs were `Variable`.
2. The second is by `Dirichlet`. Another broadcasting problem.

**Comment by fritzo** [comment]
Feel free to change the `Normal` test example from `Tensor` to `Variable`. Let me know if you want me to take a look at any test errors.

**Comment by vishwakftw** [comment]
I think I need help with the `Dirichlet` test errors. I have some difficulty solving it. Could you please help?

**Comment by fritzo** [comment]
This looks good. Could you also
1. add a docstring to the `ExponentialFamily` class that describes its role in testing and
2. add an `ExponentialFamily` section to docs/source/distributions.rst

**Comment by fritzo** [comment]
cc @martinjankowiak

File: test/test_distributions.py [hunk 4]
```diff
@@ -1857,6 +1859,23 @@ def test_kl_monte_carlo(self):
...
+                expected = ExponentialFamily.kl_divergence(p, q)
+                if isinstance(expected, Variable) and not isinstance(actual, Variable):
+                    expected = expected.data
+                self.assertEqual(actual, expected, message='\n'.join([
+                    'Incorrect KL({}, {}).'.format(type(p).__name__, type(q).__name__),
+                    'Expected (using Bregman Divergence) {}'.format(expected),
+                    'Actual (analytic) {}'.format(actual),
+                    'max error = {}'.format(torch.abs(actual - expected).max())
+                ]))
+            else:
```
Comment (fritzo) [inline test/test_distributions.py]: nit: There's no need for an `else: continue`, you can simply delete it.

File: torch/distributions/binomial.py [hunk 5]
```diff
@@ -1,96 +1,111 @@
...
-                value * self.logits + self.total_count * max_val -
-                self.total_count * torch.log1p((self.logits + 2 * max_val).exp()))
-
-    def enumerate_support(self):
-        values = self._new((self.total_count,))
-        torch.arange(self.total_count, out=values.data if isinstance(values, Variable) else values)
-        values = values.view((-1,) + (1,) * len(self._batch_shape))
-        values = values.expand((-1,) + self._batch_shape)
-        return values
+from numbers import Number
```
Comment (fritzo) [inline torch/distributions/binomial.py]: Did you change line ending format in this file? Git seems to think it changed encoding.

File: torch/distributions/exp_family.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    ExponentialFamily is the abstract base class for probability distributions belonging to an
+    exponential family, whose form is defined below
+
+    ..math::
+
+        p_{F}(x; \theta) = \exp(\langle t(x), \theta\rangle) - F(\theta) + k(x))
+
+    where :math:`\theta` denotes the natural parameters, :math:`t(x)` denotes the sufficient statistic,
+    :math:`F(\theta)` is the log normalizer function for a given family and :math:`k(x)` is the carrier
+    measure.
```
Comment (fritzo) [inline torch/distributions/exp_family.py]: Could you please add:
1. A note that this is mainly used to test correctness of `.entropy()` and `kl_divergence()` and
2. A reference to the Bregman divergence paper?

File: torch/distributions/beta.py [hunk 7]
```diff
@@ -28,6 +28,7 @@ class Beta(Distribution):
     params = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
     support = constraints.unit_interval
     has_rsample = True
+    mean_carrier_measure = float('nan')  # There is no closed form solution
```
Comment (fritzo) [inline torch/distributions/beta.py]: I think `NotImplementedError` would be more appropriate, and as you've implemented `ExponentialFamily`, simply omitting this attribute should correctly result in a `NotImplementedError`
```py
class ExponentialFamily(Distribution):
    @property
    def mean_carrier_measure(self):
        raise NotImplementedError
```

File: torch/distributions/binomial.py [hunk 8]
```diff
@@ -33,6 +33,7 @@ class Binomial(Distribution):
     """
     params = {'probs': constraints.unit_interval}
     has_enumerate_support = True
+    mean_carrier_measure = float('nan')  # There is no closed form solution
```
Comment (fritzo) [inline torch/distributions/binomial.py]: ditto: not implemented?

File: torch/distributions/dirichlet.py [hunk 9]
```diff
@@ -55,6 +55,7 @@ class Dirichlet(Distribution):
     params = {'concentration': constraints.positive}
     support = constraints.simplex
     has_rsample = True
+    mean_carrier_measure = float('nan')  # There is no closed form solution
```
Comment (fritzo) [inline torch/distributions/dirichlet.py]: ditto: leave not implemented

**Comment by alicanb** [comment]
@fritzo @vishwakftw yep, I've been watching this 👍 I don't have any request, I tried a few tricks for `Laplace`, but it I don't think it's possible for unknown `loc` and `scale`

File: torch/distributions/beta.py (hunk 7 above)
Comment (vishwakftw) [reply torch/distributions/beta.py]: Won't this break the entropy calculation? Should I add something like this?
```python
try:
    ...all entropy calculation code...
    result += self.mean_carrier_measure
except NotImplementedError:
    raise ValueError("There is no closed form solution for the entropy for non-constant carrier measure")
```

Comment (vishwakftw) [reply torch/distributions/beta.py]: I think I might have fixed it differently.

**Comment by vishwakftw** [comment]
@alicanb I managed to make it work for Binomial and Pareto, both of which had only only natural parameter. For some reason, Laplace didn't work.

File: torch/distributions/beta.py (hunk 7 above)
Comment (fritzo) [reply torch/distributions/beta.py]: I think the right behavior for `.entropy()` is to raise `NotImplementedError`, rather than return NANs. In particular, it's common in gradient estimators to e.g.
```py
try:
    entropy_estimate = dist.entropy()
except NotImplementedError:
    entropy_estimate = -dist_.log_prob(dist.sample(sample_shape=(1000,))).mean(-1)
```
If you're worried about the tests failing, I would simply wrap that line of the test in a try-except:
```
try:
    expected = dist.entropy()
except NotImplementedError:
    continue
```

Comment (vishwakftw) [reply torch/distributions/beta.py]: I think I fixed that. Please have a look at the latest push.

Comment (fritzo) [reply torch/distributions/beta.py]: You should really raise a `NotImplementedError` rather than a `ValueError`. The former means "the programmer has not implemented this; try another method", whereas the latter means "the user has input faulty data; fix your data".

File: torch/distributions/binomial.py [hunk 10]
```diff
@@ -94,3 +94,17 @@ def enumerate_support(self):
         values = values.view((-1,) + (1,) * len(self._batch_shape))
         values = values.expand((-1,) + self._batch_shape)
         return values
+
+    def entropy(self):
```
Comment (fritzo) [inline torch/distributions/binomial.py]: It looks like your default implementation of `ExponentialFamily.entropy()` already correctly raises `NotImplementedError` without performing any extra computation. Do you have a reason for adding this method here?

Comment (vishwakftw) [reply torch/distributions/binomial.py]: Oh, I have no idea why that went there. I will remove it right away. Thanks!!

**Comment by vishwakftw** [comment]
Should I close and send upstream? @fritzo

**Comment by alicanb** [comment]
@fritzo once this and transforms PR get merged, we can think about optimizations in `TransformedDistribution`s with `ExponentialFamily` base distribution.

**Comment by fritzo** [comment]
Ready to send upstream!

**Comment by ragulpr** [comment]
I really like having:
```
    @property
    def _natural_params(self):
        """
        Abstract method for natural parameters. Returns a tuple of Variables based
        on the distribution
        """
        raise NotImplementedError
```
I'm currently being confused by how to use `params` (as in `  params = {'scale': constraints.positive, 'shape': constraints.positive}`). 
If I need to infer the number of parameters it's not very nice to infer it by hacks like `len(params)`, as for multivariate or mixed distribution containing multiple `scales` one would have to explicitly write out  
`params = {'scale0': constraints.positive,....` 

But building our own schemes by inferring things from `._natural_params()` decouples this nicely. It also removes the necessity for explicit naming of distribuion parameters and adding them as `@propery`

Is `._natural_params()` something  we should implement for other distributions?

**Comment by fritzo** [comment]
I think `._natural_params()` should only be for internal use. It is not very useful for distribution-generic programming because it only applies to exponential family distributions.

@ragulpr You've clearly pointed out a need for programmatic access to params, and it sounds like this deserves an additional property, maybe a `.canonical_params` that is an `OrderedDict` from names to their current values. Note that `.params` currently over-specifies possible parameters and does not serve the purpose you want.
